### PR TITLE
Add a simplified way for eras to modify the whole process

### PR DESCRIPTION
**Pull request by mark-grimes**
**Current code**
Currently the only way for an era to make a modification to the whole process object is to create a ProcessModifier and assign it a unique name, e.g.:

``` python
def _modificationFunction( theProcess ) :
    theProcess.someThing = blah
someUniqueName_ = eras.run2_common.makeProcessModifier( _modificationFunction )
```

When the process object scans the imported objects at the end of the `load(...)` it sees that `someUniqueName_` is an instance of ProcessModifier and applies the change to itself.

This has the disadvantage of requiring a unique name, and also that it can be easily broken depending on how the file with the era change is imported.  For example, if the file above is imported with

``` python
from SomePackage.SubPackage.file import somePSet
```

Then only `somePSet` is imported and not the ProcessModifier object, hence the era changes are never applied.

**Proposal**
This pull request adds the method `toModifyProcess` to the era which stores the ProcessModifier object inside the era itself.  When the process object has finished the `load(...)` it scans all its active eras and applies any stored ProcessModifiers, then clears the list ready for the next `load(...)`.  This bypasses the import problem above and also (I think) looks much cleaner.  The new way of modifying the process object would be:

``` python
def _modificationFunction( theProcess ) :
    theProcess.someThing = blah
eras.run2_common.toModifyProcess( _modificationFunction )
```

The old way of modifying the process is still available.  If this PR is accepted then we could mark it as deprecated or remove it later.

@Dr15Jones let me know if you can see any disadvantages or have any comments.  Note that as well as adding methods I've made use of the `__processModifiers` attribute which doesn't appear to be used elsewhere.  If this is a problem I can easily modify it to use a different attribute.

### CODE REVIEW

**Comment by Dr15Jones** [comment]
Mark, I like the idea. I do have concerns about what happens if the era.toModifyProcess is used when the module in question isn't brought into python during a call to 'Process.load'.

```
from 'some.package' import 'something' #this has a teras.someEra.toModifyProcess call
from Configuration.StandardSequences.Eras import eras

process = cms.Process("Test", eras.someEra)
process.something = something
```

Under this condition the ```era.toModifyProcess``` would never be called. Now if the job later did a ```Process.load``` on some other module it would run.

Also, the case you are trying to deal with ```from * import *``` in a different python module. Under that condition, you could then only have some of the python objects to which the ```toModifyProcess``` functions operates upon. E.g. say the ```toModifyProcess``` has a line like ```del process.foo``` but via the use of ```from X import Y``` the ```foo``` is not added to the Process.

So to some extend, what does it mean to have a ```toModifyProcess``` in a python module where only part of the module is actually loaded?

**Comment by Dr15Jones** [comment]
please test

**Comment by mark-grimes** [comment]
For the first point, yes this is an issue I hadn't thought of although with the current way the modifier wouldn't be applied either.  Even if it was imported (with `...import *` or whatever) it would just be sitting in the global namespace and not executed.  I can't think of a clean solution for either way.  If your worry is that it would be executed on the next load(...) instead of not at all (would give behaviour very hard to debug), that can be fixed by wiping the store before starting the load(...) internal code.

For the second point, this also applies to the current way doesn't it?  That is, assuming the ProcessModifier was added to the import statement; if it's not then it won't be executed even if it applies to the object being imported.

In short modifying the process object from a lower-level config file is pretty messy however you go about it - I can't think of a way that covers all edge cases.

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
If we said that one can't use ```from X import Y``` on a python module containing a ```toModifyProcess``` then by combining your approach and the original approach we'd be able to tell when that occurred.

**Comment by mark-grimes** [comment]
Pretty sure the test failure "`There was a problem while querying DAS with query...`" is not down to this PR.

@Dr15Jones are you suggesting some system where both ways are used for all changes? E.g. the current way for the actual customisation and something similar to this proposal to check the first step is done?

**Comment by degano** [comment]
please test

**Comment by Dr15Jones** [comment]
@mark-grimes I am suggesting using both cases. My understanding is your original motivation for this was someone did a 'from X import Y' on a module that contained the modifier and therefore the modifier wasn't applied. So your fix would force that modify to be run. However, my concern is if some did 'from X import Y' but there was also a 'Z' in module 'X' then it is very likely the modifier will expect 'Z' to be in the Process. 
To cover both bases I was thinking of enforcing a rule that states 'if you use items from a python module that contains a process wide modifier you must import the entire module into the process.' Using both techniques together should give that guarantee.

**Comment by Dr15Jones** [comment]
@mark-grimes @davidlange6 how do you wish to proceed with this feature?

**Comment by mark-grimes** [comment]
@Dr15Jones, perhaps I misunderstood your last comment.  I thought you were saying the PR is okay as is but that we should enforce a *policy* of importing all modules.  I don't see how having both/one technique guarantees that though, so it would have to be enforced during code review or whatever.  I'm happy with the PR as is but can change it as you see fit.

**Comment by Dr15Jones** [comment]
-1
